src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java

```diff
@@ -46,14 +59,103 @@ private boolean checkIfProductsCacheNeedsUpdate() {
+        return getSearchResultsOrAllProducts(result);
+    }
+
+    private ProductList getSearchResultsOrAllProducts(List<Product> result) {
+        searchService.setSearchFailure(result.isEmpty());
+        return result.isEmpty()
+                ? new ProductList(getProductListCache().getProducts())
+                : new ProductList(result);
+    }
+
+    private void setAppliedFiltersToSearchModel(SearchModel searchModel) {
+        searchModel.setAppliedFiltersHeader(" with the applied filters");
+        String searchRequest = (null != searchService.getSearchModel().getSearchRequest())
+                ? searchService.getSearchModel().getSearchRequest()
+                : "";
+        searchModel.setSearchRequest(searchRequest);
+        searchModel.setSearchFailure(false);
+        searchModel.setActiveFilters(true);
+    }
+
+    private Predicate<Product> applyPriceFilter(SearchModel searchModel) {
+        return product -> product.getPrice().compareTo(searchModel.getPriceLow()) >= 0 && product.getPrice().compareTo(searchModel.getPriceHigh()) <= 0;
+    }
+
+    @Override
+    public ProductList sortListByName(List<Product> products, SearchModel searchModel) {
+        final Comparator<Product> productNameComparator = (searchModel.isSortAscendingName())
+                ? Comparator.comparing(Product::getName)
+                : Comparator.comparing(Product::getName).reversed();
+        return getSortedList(products, productNameComparator);
+    }
+
+    @Override
+    public ProductList sortListByPrice(List<Product> products, SearchModel searchModel) {
+        final Comparator<Product> productPriceComparator = (searchModel.isSortAscendingPrice())
+                ? Comparator.comparing(Product::getPrice)
+                : Comparator.comparing(Product::getPrice).reversed();
+        return getSortedList(products, productPriceComparator);
+    }
+
+    private ProductList getSortedList(List<Product> products, Comparator<Product> productComparator) {
+        return new ProductList(
+                products
+                        .stream()
+                        .sorted(productComparator)
+                        .collect(Collectors.toUnmodifiableList()));
+    }
+
+    private List<Product> executeSearch(String searchRequest) {
+        return (null != searchRequest)
+                ? getProductListCache().getProducts()
+                .parallelStream()
                 .filter(product -> product.getName()
                         .toLowerCase()
                         .contains(searchRequest.toLowerCase()))
-                .collect(Collectors.toUnmodifiableList());
-        return result.isEmpty() ? new ProductList(getProductListCache().getProducts()) : new ProductList(result);
+                .collect(Collectors.toUnmodifiableList())
+                : new ArrayList<>();
+    }
+
+    private List<Product> executeDescriptionSearch(List<Product> products, String description) {
+        return (null != description)
```

Comment:
boolean hasDescription(description) is more readable in this if statement.

Reply:
see explanation for the searchRequest one...